File: kotlin/src/main/kotlin/aethernet/bittorrent/Metainfo.kt
// SPDX-License-Identifier: MIT

package aethernet.bittorrent

import java.security.MessageDigest

/** One file within a torrent: its path components and length. */
data class TorrentFileEntry(val path: List<String>, val length: Long) {
    /** The path components joined with '/'. */
    fun joinedPath(): String = path.joinToString("/")
}

/**
 * A parsed BitTorrent v1 metainfo (.torrent). [infoHashV1] is the SHA-1 of the RAW
 * bencoded info dictionary as it appears in the file (extracted by byte offset, NOT
 * a re-encode), so it matches real clients byte-for-byte. The Kotlin port of the Go
 * `go/bittorrent/metainfo.go` and the C# `AetherNet.BitTorrent.Metainfo`.
 */
class TorrentMetainfo(
    val root: BDict,
    val info: BDict,
    val infoHashV1: ByteArray,
    val name: String,
    val pieceLength: Long,
    val pieceHashes: List<ByteArray>,
    val files: List<TorrentFileEntry>,
    val totalLength: Long,
    val announceUrls: List<String>,
    val isSingleFile: Boolean,
) {
    /** The lowercase hex of [infoHashV1] (40 chars). */
    fun infoHashV1Hex(): String = toHexLower(infoHashV1)
}

private fun sha1(data: ByteArray): ByteArray =
    MessageDigest.getInstance("SHA-1").digest(data)

private fun toHexLower(b: ByteArray): String {
    val sb = StringBuilder(b.size * 2)
    for (x in b) {
        val v = x.toInt() and 0xff
        sb.append("0123456789abcdef"[v ushr 4])
        sb.append("0123456789abcdef"[v and 0xf])
    }
    return sb.toString()
}

/**
 * Creates single-file .torrent bytes for [data], splitting into [pieceLength]-byte
 * pieces and SHA-1-hashing each. Byte-identical to the reference TorrentBuilder.
 */
fun buildSingleFileTorrent(name: String, data: ByteArray, pieceLength: Int, announce: String): ByteArray {
    require(name.isNotEmpty()) { "name is required" }
    require(pieceLength > 0) { "piece length must be positive" }
    val pieceCount = (data.size + pieceLength - 1) / pieceLength
    val pieces = ByteArray(pieceCount * 20)
    for (i in 0 until pieceCount) {
        val start = i * pieceLength
        val end = minOf(start + pieceLength, data.size)
        val h = sha1(data.copyOfRange(start, end))
        System.arraycopy(h, 0, pieces, i * 20, 20)
    }

    val info = BDict()
    info.add("length", BInt(data.size.toLong()))
    info.add("name", BStr(name))
    info.add("piece length", BInt(pieceLength.toLong()))
    info.add("pieces", BStr(pieces))

    val root = BDict()
    if (announce.isNotBlank()) root.add("announce", BStr(announce))
    root.add("info", info)
    return Bencode.encode(root)
}

/** Parses .torrent bytes into a [TorrentMetainfo]. */
fun parseTorrent(data: ByteArray): TorrentMetainfo {
    val root = Bencode.decode(data).asDict()
    val info = (root.get("info") ?: throw BencodeException("metainfo has no 'info' dictionary")).asDict()

    val infoSpan = extractInfoSpan(data)
    val infoHash = sha1(infoSpan)

    val name = (info.get("name") ?: throw BencodeException("info has no 'name'")).asText()

    val pieceLength = (info.get("piece length") ?: throw BencodeException("info has no 'piece length'")).asInt()
    if (pieceLength <= 0) throw BencodeException("'piece length' must be positive")

    val piecesBytes = (info.get("pieces") ?: throw BencodeException("info has no 'pieces'")).asBytes()
    if (piecesBytes.size % 20 != 0) {
        throw BencodeException("'pieces' length ${piecesBytes.size} is not a multiple of 20")
    }
    val pieceHashes = ArrayList<ByteArray>(piecesBytes.size / 20)
    var i = 0
    while (i < piecesBytes.size) {
        pieceHashes.add(piecesBytes.copyOfRange(i, i + 20))
        i += 20
    }

    val files = ArrayList<TorrentFileEntry>()
    var total = 0L
    var singleFile = false
    val filesVal = info.get("files")
    if (filesVal != null) {
        for (f in filesVal.asList()) {
            val fd = f.asDict()
            val length = (fd.get("length") ?: throw BencodeException("file entry has no 'length'")).asInt()
            val pathList = (fd.get("path") ?: throw BencodeException("file entry has no 'path'")).asList()
            val parts = pathList.map { it.asText() }
            if (parts.isEmpty()) throw BencodeException("file entry has an empty 'path'")
            files.add(TorrentFileEntry(parts, length))
            total += length
        }
    } else {
        singleFile = true
        val length = (info.get("length")
            ?: throw BencodeException("single-file info has neither 'length' nor 'files'")).asInt()
        files.add(TorrentFileEntry(listOf(name), length))
        total = length
    }

    // Trackers: announce + announce-list, de-duplicated, order preserved.
    val announce = ArrayList<String>()
    val seen = HashSet<String>()
    fun add(u: String) {
        if (u.isNotEmpty() && seen.add(u)) announce.add(u)
    }
    (root.get("announce") as? BStr)?.let { add(it.asText()) }
    (root.get("announce-list") as? BList)?.let { al ->
        for (tier in al.items) {
            (tier as? BList)?.let { ts ->
                for (t in ts.items) (t as? BStr)?.let { add(it.asText()) }
            }
        }
    }

    return TorrentMetainfo(
        root = root,
        info = info,
        infoHashV1 = infoHash,
        name = name,
        pieceLength = pieceLength,
        pieceHashes = pieceHashes,
        files = files,
        totalLength = total,
        announceUrls = announce,
        isSingleFile = singleFile,
    )
}

/**
 * Returns the raw bencoded bytes of the top-level "info" value by walking the
 * dictionary with byte-offset tracking (structure already validated by [parseTorrent]).
 */
internal fun extractInfoSpan(data: ByteArray): ByteArray {
    if (data.isEmpty() || (data[0].toInt() and 0xff) != 'd'.code) {
        throw BencodeException("metainfo is not a bencoded dictionary")
    }
    val r = BencodeReader(data)
    r.pos = 1
    while (r.pos < data.size && (data[r.pos].toInt() and 0xff) != 'e'.code) {
        val key = r.readValue().asBytes()
        val valStart = r.pos
        r.readValue()
        val valEnd = r.pos
        if (String(key, Charsets.UTF_8) == "info") return data.copyOfRange(valStart, valEnd)
    }
    throw BencodeException("metainfo has no 'info' key")
}

File: python/aethernet/bittorrent/metainfo.py
# SPDX-License-Identifier: MIT
"""Torrent metainfo, info-hash (SHA-1 of the raw info dict), magnet, and builder."""
from __future__ import annotations

import base64
import hashlib
from urllib.parse import parse_qs, urlsplit

from . import bencode


class TorrentError(ValueError):
    pass


class TorrentMetainfo:
    def __init__(self, root, info, info_hash_v1, name, piece_length, piece_hashes,
                 files, total_length, announce_urls, is_single_file):
        self.root = root
        self.info = info
        self.info_hash_v1 = info_hash_v1
        self.name = name
        self.piece_length = piece_length
        self.piece_hashes = piece_hashes
        self.files = files
        self.total_length = total_length
        self.announce_urls = announce_urls
        self.is_single_file = is_single_file

    @property
    def info_hash_v1_hex(self) -> str:
        return self.info_hash_v1.hex()


def build_single_file_torrent(name: str, data: bytes, piece_length: int, announce: str = "") -> bytes:
    if not name:
        raise TorrentError("name is required")
    if piece_length <= 0:
        raise TorrentError("piece length must be positive")
    piece_count = (len(data) + piece_length - 1) // piece_length
    pieces = bytearray()
    for i in range(piece_count):
        start = i * piece_length
        end = min(start + piece_length, len(data))
        pieces += hashlib.sha1(data[start:end]).digest()
    info = {
        b"length": len(data),
        b"name": name.encode("utf-8"),
        b"piece length": piece_length,
        b"pieces": bytes(pieces),
    }
    root: dict = {}
    if announce and announce.strip():
        root[b"announce"] = announce.encode("utf-8")
    root[b"info"] = info
    return bencode.encode(root)


def parse_torrent(data: bytes) -> TorrentMetainfo:
    root = bencode.decode(data)
    if not isinstance(root, dict):
        raise TorrentError("metainfo is not a dictionary")
    info = root.get(b"info")
    if not isinstance(info, dict):
        raise TorrentError("metainfo has no 'info' dictionary")

    info_hash = hashlib.sha1(_extract_info_span(data)).digest()

    name = _text(info.get(b"name"), "info has no 'name'")
    piece_length = _int(info.get(b"piece length"), "info has no 'piece length'")
    if piece_length <= 0:
        raise TorrentError("'piece length' must be positive")

    pieces = info.get(b"pieces")
    if not isinstance(pieces, (bytes, bytearray)):
        raise TorrentError("info has no 'pieces'")
    if len(pieces) % 20 != 0:
        raise TorrentError("'pieces' length is not a multiple of 20")
    piece_hashes = [bytes(pieces[i:i + 20]) for i in range(0, len(pieces), 20)]

    files = []
    total = 0
    if b"files" in info:
        is_single = False
        for f in info[b"files"]:
            length = _int(f.get(b"length"), "file entry has no 'length'")
            parts = [p.decode("utf-8") for p in f.get(b"path", [])]
            if not parts:
                raise TorrentError("file entry has an empty 'path'")
            files.append((parts, length))
            total += length
    else:
        is_single = True
        length = _int(info.get(b"length"), "single-file info has neither 'length' nor 'files'")
        files.append(([name], length))
        total = length

    announce = []
    seen = set()

    def add(u: str):
        if u and u not in seen:
            seen.add(u)
            announce.append(u)

    if b"announce" in root:
        add(root[b"announce"].decode("utf-8"))
    for tier in root.get(b"announce-list", []):
        for t in tier:
            add(t.decode("utf-8"))

    return TorrentMetainfo(root, info, info_hash, name, piece_length, piece_hashes,
                           files, total, announce, is_single)


def _extract_info_span(data: bytes) -> bytes:
    if not data or data[0] != ord("d"):
        raise TorrentError("metainfo is not a bencoded dictionary")
    pos = 1
    while pos < len(data) and data[pos] != ord("e"):
        key, pos = bencode._decode_str(data, pos)
        val_start = pos
        _, pos = bencode.decode_n(data, pos)
        if key == b"info":
            return data[val_start:pos]
    raise TorrentError("metainfo has no 'info' key")


def _text(v, err: str) -> str:
    if not isinstance(v, (bytes, bytearray)):
        raise TorrentError(err)
    return bytes(v).decode("utf-8")


def _int(v, err: str) -> int:
    if not isinstance(v, int):
        raise TorrentError(err)
    return v


class MagnetLink:
    def __init__(self, info_hash: bytes, display_name: str, trackers: list[str]):
        self.info_hash = info_hash
        self.display_name = display_name
        self.trackers = trackers

    @property
    def info_hash_hex(self) -> str:
        return self.info_hash.hex()


def parse_magnet(uri: str) -> MagnetLink:
    if not uri.startswith("magnet:?"):
        raise TorrentError("not a magnet URI")
    q = parse_qs(uri[len("magnet:?"):])
    info_hash = None
    for xt in q.get("xt", []):
        if xt.startswith("urn:btih:"):
            info_hash = _decode_info_hash(xt[len("urn:btih:"):])
            break
    if info_hash is None:
        raise TorrentError("magnet has no xt=urn:btih: topic")
    dn = q.get("dn", [""])[0]
    return MagnetLink(info_hash, dn, q.get("tr", []))


def _decode_info_hash(s: str) -> bytes:
    if len(s) == 40:
        return bytes.fromhex(s)
    if len(s) == 32:
        return base64.b32decode(s.upper())
    raise TorrentError(f"info-hash must be 40 hex or 32 base32 chars, got {len(s)}")
